py_proxy/views.py

```diff
@@ -38,7 +38,9 @@ def favicon(request):
 def pdf(request):
     """HTML page with client and the PDF embedded."""
     nginx_server = request.registry.settings["nginx_server"]
-    pdf_url = request.matchdict["pdf_url"]
+    # Drop /pdf/ from url.
+    pdf_url = request.path_qs[5:]
```

Comment:
Can you add a unit test for the query string preserving here? There doesn't seem to be a test that ensures that the query string is preserved on `pdf_url`. There should be tests for both when the request does and does not have a query string.

I think adding a unit test is the main thing but the code could be _slightly_ nicer by removing the hard-coded `5` in one way or another. Here are a couple of suggestions that I think would work:

You can get the PDF URL (without query string) as `request.matchdict['pdf_url']` and then get the query string (without the leading `?`) as `request.query_string`, so you can construct the desired `pdf_url` without indexing and without `urllib3` like this:

```suggestion
    pdf_url = request.matchdict["pdf_url"]
    if request.query_string:
        pdf_url = pdf_url + "?" + request.query_string
```

This is longer but it's more explicit and readable, and the code is more decoupled from the exact URL path string.

Or alternatively use `len("/pdf/")` to make it more explicit that you're truncating the `"/pdf/"` from the beginning of the string:

```python
pdf_url = request.path_qs[len("/pdf/"):]
```

Reply:
I'm going with the len("/pdf/") approach and I've made it into a private helper function as it's used in a couple places.